streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala

```diff
@@ -361,6 +392,105 @@ class ZStream[-R, +E, +A](val channel: ZChannel[R, Any, Any, Any, E, Chunk[A], A
     )
   }
 
+  /**
+   * Allows a faster producer to progress independently of a slower consumer by buffering
+   * up to `capacity` elements in a dropping queue.
+   *
+   * @note Prefer capacities that are powers of 2 for better performance.
+   */
+  final def bufferDropping(capacity: Int): ZStream[R, E, A] = {
+    val queue = Queue.dropping[(Take[E, A], Promise[Nothing, Unit])](capacity).toManaged(_.shutdown)
+    new ZStream(bufferSignal[R, E, A](queue, _.pullElement.map(Chunk.single)))
+  }
+
+  /**
+   * Allows a faster producer to progress independently of a slower consumer by buffering
+   * up to `capacity` elements in a sliding queue.
+   *
+   * @note Prefer capacities that are powers of 2 for better performance.
+   */
+  final def bufferSliding(capacity: Int): ZStream[R, E, A] = {
+    val queue = Queue.sliding[(Take[E, A], Promise[Nothing, Unit])](capacity).toManaged(_.shutdown)
+    new ZStream(bufferSignal[R, E, A](queue, _.pullElement.map(Chunk.single)))
+  }
+
+  private def bufferSignal[R1 <: R, E1 >: E, A1 >: A](
+    managed: UManaged[Queue[(Take[E1, A1], Promise[Nothing, Unit])]],
+    toPull: BufferedPull[R1, E1, A1] => ZStream.Pull[R1, E1, A1]
+  ): ZChannel[R1, Any, Any, Any, E1, Chunk[A1], Unit] = {
+    def runProducer(queue: Queue[(Take[E1, A1], Promise[Nothing, Unit])]): ZManaged[R1, Nothing, Unit] = {
+      def offer(ref: Ref[Promise[Nothing, Unit]])(take: Take[E1, A1]): UIO[Unit] =
+        take.exit.fold(
+          _ =>
+            for {
+              latch <- ref.get
+              _     <- latch.await
+              p     <- Promise.make[Nothing, Unit]
+              _     <- queue.offer((take, p))
+              _     <- ref.set(p)
+              _     <- p.await
+            } yield (),
+          _ =>
+            for {
+              p     <- Promise.make[Nothing, Unit]
+              added <- queue.offer((take, p))
+              _     <- ref.set(p).when(added)
+            } yield ()
+        )
+
+      for {
+        as    <- self.toPull.mapM(BufferedPull.make[R1, E1, A1](_))
+        start <- Promise.make[Nothing, Unit].toManaged_
+        _     <- start.succeed(()).toManaged_
+        ref   <- Ref.makeManaged(start)
+        _     <- Take.fromPull(toPull(as)).tap(offer(ref)).repeatWhile(_.isSuccess).unit.toManaged_.fork
+      } yield ()
+    }
```

Comment:
It seems that we can write this without `toPull` as we are not relying on reading from it in another channel. Is that possible?

Reply:
Another interesting note is that @adamgfraser has shown that we can write things that use sliding buffers using the ring buffer directly. Would you like to look into that here?

Reply:
I will try now to implement this part without `toPull`. I just rewrote `runIntoElementsManaged` without it. Is there any place where you can point me at to look at the ring buffer?

Reply:
Re implemented without Pull. Maybe explore the ring buffer in a follow-up @iravid?

Reply:
I just found Adam's implementation of the ring buffer. I will update this PR with its usage.